File: push/src/main/java/com/xiaomi/push/service/JobSchedulerThread.java
package com.xiaomi.push.service;

import android.os.SystemClock;

final class JobSchedulerThread extends Thread {
    private static final int MAX_SLEEP_TIME = 500;
    private static final int SLEEP_TIME_STEP = 50;
    private boolean cancelled;
    private boolean finished;
    private volatile long lastJob = 0;
    private volatile boolean executing = false;
    private long currentSleepDuration = 50;
    private final JobSchedulerTaskQueue tasks = new JobSchedulerTaskQueue();

    JobSchedulerThread(String str, boolean z) {
        setName(str);
        setDaemon(z);
        start();
    }

    void insertTask(JobSchedulerTaskWrapper jobSchedulerTaskWrapper) {
        this.tasks.insert(jobSchedulerTaskWrapper);
        notify();
    }

    void cancelScheduler() {
        synchronized (this) {
            this.cancelled = true;
            this.tasks.reset();
            notify();
        }
    }

    boolean isBlocked() {
        return this.executing && SystemClock.uptimeMillis() - this.lastJob > 600000;
    }

    int purge() {
        return this.tasks.purge();
    }

    JobSchedulerTaskQueue tasks() {
        return this.tasks;
    }

    void finish() {
        synchronized (this) {
            this.finished = true;
            notify();
        }
    }

    boolean isCancelled() {
        return this.cancelled;
    }

    @Override
    public void run() {
        while (true) {
            try {
                JobSchedulerTaskWrapper jobSchedulerTaskWrapper = null;
                synchronized (this) {
                    if (this.cancelled) {
                        return;
                    }
                    if (this.tasks.isEmpty()) {
                        if (this.finished) {
                            return;
                        }
                        try {
                            wait();
                        } catch (InterruptedException e) {
                        }
                        continue;
                    }
                    long currentTime = JobSchedulerClock.getCurrentTime();
                    jobSchedulerTaskWrapper = this.tasks.minimum();
                    synchronized (jobSchedulerTaskWrapper.lock) {
                        if (jobSchedulerTaskWrapper.cancelled) {
                            this.tasks.delete(0);
                            continue;
                        }
                        long j = jobSchedulerTaskWrapper.when - currentTime;
                        if (j > 0) {
                            long j2 = this.currentSleepDuration;
                            if (j > j2) {
                                j = j2;
                            }
                            long j3 = j2 + SLEEP_TIME_STEP;
                            this.currentSleepDuration = j3;
                            if (j3 > MAX_SLEEP_TIME) {
                                this.currentSleepDuration = MAX_SLEEP_TIME;
                            }
                            try {
                                wait(j);
                            } catch (InterruptedException e2) {
                            }
                            continue;
                        }
                        this.currentSleepDuration = SLEEP_TIME_STEP;
                        int task = 0;
                        if (this.tasks.minimum().when != jobSchedulerTaskWrapper.when) {
                            task = this.tasks.getTask(jobSchedulerTaskWrapper);
                        }
                        if (jobSchedulerTaskWrapper.cancelled) {
                            this.tasks.delete(this.tasks.getTask(jobSchedulerTaskWrapper));
                            continue;
                        }
                        jobSchedulerTaskWrapper.setScheduledTime(jobSchedulerTaskWrapper.when);
                        this.tasks.delete(task);
                        jobSchedulerTaskWrapper.when = 0L;
                    }
                }
                this.lastJob = SystemClock.uptimeMillis();
                this.executing = true;
                jobSchedulerTaskWrapper.job.run();
                this.executing = false;
            } catch (Throwable th) {
                synchronized (this) {
                    this.cancelled = true;
                }
                throw th;
            }
        }
    }
}

File: legacy-runtime/src/main/java/com/xiaomi/push/service/JobSchedulerThread.kt
package com.xiaomi.push.service

import android.os.SystemClock

class JobSchedulerThread(
    name: String,
    daemon: Boolean,
) : Thread() {
    companion object {
        private const val MAX_SLEEP_TIME = 500L
        private const val SLEEP_TIME_STEP = 50L
        private const val BLOCKED_THRESHOLD_MS = 600000L
    }

    private var cancelled = false
    private var finished = false

    @Volatile
    private var lastJob = 0L

    @Volatile
    private var executing = false

    private var currentSleepDuration = SLEEP_TIME_STEP
    private val taskQueue = JobSchedulerTaskQueue()

    private fun signalLocked() {
        @Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")
        (this as java.lang.Object).notify()
    }

    private fun waitLocked(timeoutMs: Long? = null) {
        @Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")
        val objectMonitor = this as java.lang.Object
        if (timeoutMs == null) {
            objectMonitor.wait()
        } else {
            objectMonitor.wait(timeoutMs)
        }
    }

    init {
        setName(name)
        isDaemon = daemon
        start()
    }

    fun insertTask(taskWrapper: JobSchedulerTaskWrapper) {
        taskQueue.insert(taskWrapper)
        signalLocked()
    }

    fun cancelScheduler() {
        synchronized(this) {
            cancelled = true
            taskQueue.reset()
            signalLocked()
        }
    }

    fun isBlocked(): Boolean {
        return executing && SystemClock.uptimeMillis() - lastJob > BLOCKED_THRESHOLD_MS
    }

    fun purge(): Int = taskQueue.purge()

    fun tasks(): JobSchedulerTaskQueue = taskQueue

    fun finish() {
        synchronized(this) {
            finished = true
            signalLocked()
        }
    }

    fun isCancelled(): Boolean = cancelled

    override fun run() {
        while (true) {
            try {
                var taskWrapper: JobSchedulerTaskWrapper? = null
                var shouldContinue = false

                synchronized(this) {
                    if (cancelled) {
                        return
                    }
                    if (taskQueue.isEmpty()) {
                        if (finished) {
                            return
                        }
                        try {
                            waitLocked()
                        } catch (_: InterruptedException) {
                        }
                        shouldContinue = true
                    } else {
                        val currentTime = JobSchedulerClock.getCurrentTime()
                        val task = taskQueue.minimum()
                        taskWrapper = task
                        synchronized(task.lock) {
                            if (task.cancelled) {
                                taskQueue.delete(0)
                                shouldContinue = true
                            } else {
                                var sleepDuration = task.`when` - currentTime
                                if (sleepDuration > 0) {
                                    if (sleepDuration > currentSleepDuration) {
                                        sleepDuration = currentSleepDuration
                                    }
                                    currentSleepDuration =
                                        minOf(currentSleepDuration + SLEEP_TIME_STEP, MAX_SLEEP_TIME)
                                    try {
                                        waitLocked(sleepDuration)
                                    } catch (_: InterruptedException) {
                                    }
                                    shouldContinue = true
                                } else {
                                    currentSleepDuration = SLEEP_TIME_STEP
                                    var taskIndex = 0
                                    if (taskQueue.minimum().`when` != task.`when`) {
                                        taskIndex = taskQueue.getTask(task)
                                    }
                                    if (task.cancelled) {
                                        taskQueue.delete(taskQueue.getTask(task))
                                        shouldContinue = true
                                    } else {
                                        task.setScheduledTime(task.`when`)
                                        taskQueue.delete(taskIndex)
                                        task.`when` = 0L
                                    }
                                }
                            }
                        }
                    }
                }

                if (shouldContinue) {
                    continue
                }

                lastJob = SystemClock.uptimeMillis()
                executing = true
                try {
                    checkNotNull(taskWrapper?.job).run()
                } finally {
                    executing = false
                }
            } catch (throwable: Throwable) {
                synchronized(this) {
                    cancelled = true
                }
                throw throwable
            }
        }
    }
}
